### PR TITLE
Jonas' article on Effection

### PR DESCRIPTION
**Pull request by jorgelainfiesta**
## Motivation

Publish new article written by Jonas here: https://github.com/thefrontside/effection/pull/575

## Approach

We moved the error example down because it was too much introduction before getting into Effection. 

I'm posting this article in frontside.com because I want to centralize our content to a single place. I hope that we can have a deeper integration from the OSS projects to the .com website in a near future.

### CODE REVIEW

File: src/blog/2021-10-26-effection-async-await.md [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+
+That&#39;s why I was struck so powerfully by something I read recentlya while ago but keep coming back to that illuminated this issue for me. It&#39;s the kind of rare piece of writing that fundamentally shifts your perspective on how things _should_ be done – where after having read it you have this overwhelming feeling of almost panic that _obviously_ the author is correct and _obviously_ this is the way to do it. That was exactly the feeling we had after reading Nathaniel J. Smith&#39;s masterful blog post [Notes on structured concurrency, or: Go statement considered harmful](https://vorpus.org/blog/notes-on-structured-concurrency-or-go-statement-considered-harmful/):
```
Comment (taras) [inline src/blog/2021-10-26-effection-async-await.md]: `recentlya while ago` is it supposed to be `recently` or `a while ago`?

File: src/blog/2021-10-26-effection-async-await.md [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+
+> Just like goto was the obvious primitive for the first practical high-level languages, go was the obvious primitive for the first practical concurrency frameworks: it matches how the underlying schedulers actually work, and it&#39;s powerful enough to implement any other concurrent flow pattern. But again like goto, it breaks control flow abstractions, so that merely having it as an option in your language makes everything harder to use.
+> 
+> The good news, though, is that these problems can all be solved: Dijkstra showed us how! We need to:
+> 
+>   - Find a replacement for go statements that has similar power, but follows the &quot;black box rule&quot;
+>   - Build that new construct into our concurrency framework as a primitive, and don&#39;t include any form of go statement.
```
Comment (taras) [inline src/blog/2021-10-26-effection-async-await.md]: Can we make this look more like a quote? It's a bit difficult to tell that it's quoting the article.

image

File: src/blog/2021-10-26-effection-async-await.md (hunk 1 above)
Comment (jorgelainfiesta) [reply src/blog/2021-10-26-effection-async-await.md]: `a while ago`, somehow the edit history got dragged in my copy+paste

File: src/blog/2021-10-26-effection-async-await.md [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+
+But what was truly remarkable about his post was not only did he describe the contours of the abyss perfectly, but he points out exactly the right route you need to follow to climb out.
+
+## Introducing Effection
+
+The Smith&#39;s pathway out of the abyss is to adopt the ideas of a structured concurrency and apply them to JavaScript. The core idea of structured concurrency is that the lifetime of a task is always constrained. This means that a task must not outlive its parent, or in plain English, it is impossible to create a task that runs _forever_. This might seem like an obvious constraint, but it is important to note that this is very much _not_ the case with the core concurrency primitives available in JavaScript.
```
Comment (cowboyd) [inline src/blog/2021-10-26-effection-async-await.md]:
```suggestion
Dijkstra&#39;s pathway out of the abyss is to adopt the ideas of a structured concurrency and apply them to JavaScript. The core idea of structured concurrency is that the lifetime of a task is always constrained. This means that a task must not outlive its parent, or in plain English, it is impossible to create a task that runs _forever_. This might seem like an obvious constraint, but it is important to note that this is very much _not_ the case with the core concurrency primitives available in JavaScript.
```

File: src/blog/2021-10-26-effection-async-await.md [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+
+We have experimented with these ideas for a while at [Frontside](http://frontside.com/effection), and as a result we&#39;ve created a framework that has changed the way we write and think about JavaScript systems. This process has been challenging and exciting, and we think that we&#39;ve built something which is useful far beyond our own team and our own needs that we&#39;d like to share with others.
+
+[Effection](http://frontside.com/effection) is our Open Source concurrency framework that replaces async/await with a more structured way of writing code. We hope that it offers a base out of which to build an entire ecosystem, but we&#39;ve also found it to be incredibly useful and fun to employ on its own. Let&#39;s look at how this applies to async/await code:
```
Comment (cowboyd) [inline src/blog/2021-10-26-effection-async-await.md]:
```suggestion
[Effection](http://frontside.com/effection) is our Open Source concurrency framework that replaces async/await with a more structured way of writing code. We hope that it offers a base off of which to build an entire ecosystem, but we&#39;ve also found it to be incredibly useful and fun to employ on its own. Let&#39;s look at how this applies to async/await code:
```

File: src/blog/2021-10-26-effection-async-await.md [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+
+```javascript
+async function fetchUser(id) {
+  let response = await fetch(`/users/${id}`);
+  return await response.json();
+}
+```
+
+There is nothing special about this async function. And this is what happens when we usecall this function from another function:
```
Comment (cowboyd) [inline src/blog/2021-10-26-effection-async-await.md]:
```suggestion
There is nothing special about this async function. And this is what happens when we call it from another function:
```

**Review by cowboyd — COMMENTED** [review]
This is great! I had the biggest smile the whole time I was reading this.

I added some small fixes, but after those, I say shipit!

**Comment by jorgelainfiesta** [comment]
Thanks @cowboyd! I've applied the typo change you found. The others changes are in paragraphs that are no longer present.

@taras I've updated the tone as requested, let me know if this works

**Review by taras — APPROVED** [review]
giphy-9


👍

File: src/blog/2021-10-26-effection-async-await.md [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+async function fetchSomeUsers() {
+  let userOne = fetchUser(1);
+  let userTwo = fetchUser(2);
+  return {
+    one: await userOne,
+    two: await userTwo,
+  }
+}
+```
+> Note: we could have used `Promise.all` for this, the resulting behaviour would be the same.
```
Comment (cowboyd) [inline src/blog/2021-10-26-effection-async-await.md]:
```suggestion
> Note: we could have used `Promise.all()` for this, but the resulting behavior would be the same.
```

File: src/blog/2021-10-26-effection-async-await.md [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+    two: await userTwo,
+  }
+}
+```
+
+What will happen in this case? `fetchUser(1)` and `fetchUser(2)` will happily keep running, even though the `fetchSomeUsers` function which initially called them has already failed.
+
+![fetchSomeUsers timing with async/await](/img/2021-effection-intro/intro-effection-async-await.svg)
+
+This can&#39;t happen in Effection. Because given that `fetchUser(1)` and `fetchUser(2)` are scoped to their parent function, they will be terminated when `fetchSomeUsers` enters in an error state.
```
Comment (cowboyd) [inline src/blog/2021-10-26-effection-async-await.md]:
```suggestion
This can&#39;t happen in Effection. Because given that `fetchUser(1)` and `fetchUser(2)` are scoped to their parent function, they will be terminated when `fetchSomeUsers` enters into an error state.
```

**Review by cowboyd — COMMENTED** [review]
Just a couple more grammatical fixes I noticed.

**Comment by jorgelainfiesta** [comment]
Thanks @cowboyd!
